ruby/phase-1.md

```diff
@@ -0,0 +1,398 @@
+
+config_2 = Temporal::Configuration.new(
+  url: 'host-2.com:7233',
+  namespace: 'namespace-2',
+  task_queue: 'queue-2'
+)
+client_2 = Temporal::Client.new(config_2)
+worker_2 = Temporal::Worker.new(config_2)
+```
+
+
+## Activity/Workflow execution attributes
+
+This is an important and novel (as far as all the other SDKs are concerned) concept that allows to
+reduce the explicitness of using the Ruby SDK. This is achieved via a few assumptions that have been
+tested against the real world use-cases, but obviously are up for a debate. These are:
+
+- The simplest __viable__ use-case can be satisfied with a single namespace and single task queue
+- Most workflow/activity names are identical to the names of the classes implementing them
+- The most popular setup will be based on a one-to-one mapping between workflows/activities and a
+  namespace + task queue combination
+- In most cases timeouts and retry policies are defined per workflow/activity and not a specific
+  execution. That is to say that workflow/activity owners will probably have the best idea of which
+  timeouts and retry policies should be used when executing it
+
+With that in mind we're proposing a way to configure the said attribute per workflow/activity
+definition. This would reside in the outer API and would look something like this:
+
+```ruby
+class NewWorkflow < Temporal::Workflow
+  namespace 'ns-1'
+  task_queue 'tq-1'
+  timeouts run: 10 * 60 * 60, execution: 20 * 60 * 60
+
+  def execute
+    ...
+  end
+end
+
+class AnotherActivity < Temporal::Activity
+  namespace 'ns-2'
+  task_queue 'tq-2'
+  retry_policy(
+    interval: 1,
+    backoff: 1,
+    max_attempts: 3,
+    non_retriable_errors: [FatalError]
+  )
+
+  def execute
+    ...
+  end
+end
+```
+
+The main idea here is when these are started/scheduled the caller doesn't have to explicitly specify
+`name`, `namespace`, `task_queue`, `timeouts` or `retry_policy`, but can instead rely on the
+existing definitions. The same exact concept applies to registering these classes with a worker,
+that can then figure out which namespace and task queue need to be polled. Here's what that looks
+like:
+
+```ruby
+# starting a workflow
+client.start_workflow(NewWorkflow) # assumes namespace: 'ns-1', task_queue: 'tq-1'
+
+# registering classes with a worker
+worker.register_workflow(NewWorkflow) # assumes namespace: 'ns-1', task_queue: 'tq-1'
+worker.register_activity(NewActivity) # assumes namespace: 'ns-2', task_queue: 'tq-2'
+```
+
+This of course can be fully or partially overwritten by the caller, which has the full control over
+the resulting attributes. For example:
+
+```ruby
+client.start_workflow(NewWorkflow, options: { namespace: 'ns-3' }) # namespace: 'ns-3', task_queue: 'tq-1'
+
+worker.register_workflow(NewWorkflow, task_queue: 'tq-3') # namespace: 'ns-1', task_queue: 'tq-3'
+worker.register_activity(NewActivity, namespace: 'ns-3', task_queue: 'tq-3') # namespace: 'ns-3', task_queue: 'tq-3'
```

Comment:
Is there any benefit in these method calls mimicking the method signature of `#start_workflow`? Specifically, #start_workflow uses an `options` hash to pass modified arguments but the #register_* methods take them as direct args. Would it be better if the #register_* methods also accepted an `options` hash instead?

I like that consistency.

On the flipside, I think the options hash should also be a first-class Struct (with compatibility via #to_h) because I don't like passing around hashes. Probably a different topic...

Reply:
These are slightly different signatures and the similarities are probably not very important. I do get the point about consistency, but it might feel a bit weird to have a single argument + `options:`. But maybe it's just me, so I'll change that

As for the Struct — very good point. One downside to this approach is that a Struct doesn't distinguish between a `nil` and a missing key, but maybe it's not that important